near-contract-standards/src/non_fungible_token/payout/payout_impl.rs

```diff
@@ -0,0 +1,323 @@
+use super::NonFungibleTokenPayout;
+use crate::fungible_token::Balance;
+use crate::non_fungible_token::core::NonFungibleTokenCore;
+use crate::non_fungible_token::payout::*;
+use crate::non_fungible_token::NonFungibleToken;
+use near_sdk::{assert_one_yocto, env};
+use near_sdk::{require, AccountId, IntoStorageKey};
+
+impl Royalties {
+    pub fn new<S>(key_prefix: S) -> Self
+    where
+        S: IntoStorageKey,
+    {
+        let temp_accounts: TreeMap<AccountId, HashMap<TokenId, BasisPoint>> =
+            TreeMap::new(key_prefix);
+        let this = Self { accounts: temp_accounts };
+        this.validate();
+        this
+    }
+
+    pub(crate) fn validate(&self) {
+        require!(
+            self.accounts.len() <= 10,
+            "can only have a maximum of 10 accounts spliting royalties"
+        );
+
+        let mut total_per_token = HashMap::new();
+
+        self.accounts.iter().for_each(|(_, percent_per_token)| {
+            percent_per_token.iter().for_each(|(token_id, percent)| {
+                require!(*percent <= 100, "each royalty should be at most 100");
+                *total_per_token.entry(token_id.to_owned()).or_default() += percent;
+            });
+        });
+
+        total_per_token.values().for_each(|total: &u16| {
+            require!(
+                *total <= 100,
+                "total percent of each royalty split must be at most 100 per token"
+            )
+        });
+    }
+
+    /// Create a payout.
+    ///
+    /// # Arguments
+    /// * `token_id` - token_id for payout.
+    /// * `balance` - total balance dedicated to the payout.
+    /// * `owner_id` - nft owner account id.
+    ///
+    /// NOTE: The owner gets whatever is left after distributing the rest of the payout plus the
+    /// percentage specified explicitly, if any.
+    pub fn create_payout(
+        &self,
+        token_id: TokenId,
+        balance: Balance,
+        owner_id: &AccountId,
+    ) -> Payout {
+        let mut payout = Payout {
+            payout: self
+                .accounts
+                .iter()
+                .map(|(account, percent_per_token)| {
+                    (
+                        account.clone(),
+                        U128::from(apply_percent(
+                            *percent_per_token.get(&token_id).unwrap_or(&0),
+                            balance,
+                        )),
+                    )
+                })
+                .filter(|(_, payout)| payout.0 > 0)
+                .collect(),
+        };
+        let rest = balance - payout.payout.values().fold(0, |acc, &sum| acc + sum.0);
+        let owner_payout: u128 = payout.payout.get(owner_id).map_or(0, |x| x.0) + rest;
+        payout.payout.insert(owner_id.clone(), owner_payout.into());
```

Comment:
Something crazy is going on here. Royalties should be just an `UnorderedMap<TokenId, HashMap<AccountId, Percentage>>`, and then `create_payout` will be just: 

```rust
royalties.get(token_id)
    .items()
    .map(|(account_id, percentage)| (account_id, balance * percentage / 100))
    .collect()
```

Reply:
@ruseinov Also, as a rule of thumb, if you write a code that iterates over near_sdk collections without bounds and potentially may exceed more than 100 lookups, you need to redesign the approach as that won't work with the growing number records. In this case, as I mentioned above, it needs a completely different layout

Reply:
My mistake, I made a fast fix of a problem that completely redefined the solution.

Reply:
Actually, the truth is that the original map is bound by `validate` method, it only allows for 10 accounts. So technically it's not an issue. 
The data structure and it's access are a little bit convoluted, but if we were to redesign this collection and the approach - we'd have to discuss how we are going to validate bounds going forward.